src/java/org/apache/cassandra/io/sstable/StorageHandler.java

```diff
@@ -0,0 +1,149 @@
+
+import static org.apache.cassandra.config.CassandraRelevantProperties.REMOTE_STORAGE_HANDLER;
+
+/**
+ * The handler of the storage of sstables, and possibly other files such as txn logs.
+ * <p/>
+ * If sstables are stored on the local disk, then this handler is a thin wrapper of {@link Directories.SSTableLister},
+ * but for sstables stored remotely, for example on S3, then the handler may need to perform more
+ * work, such as selecting only part of the remote sstables available, or adding new ones when offline compaction
+ * has run. This behaviour can be implemented in a sub-class that can be set with {@link #remoteStorageHandler}.
+ * <p/>
+ * Local sytem tables will always use the default storage handler, {@link DefaultStorageHandler}
+ * whereas user tables will use {@link #remoteStorageHandler} if one is set, or the default handler if none
+ * has been set.
+ * <p/>
+ */
+public abstract class StorageHandler
+{
+    private final static String remoteStorageHandler = REMOTE_STORAGE_HANDLER.getString();
+
+    public enum ReloadReason
+    {
+        /** New nodes joined or left */
+        TOPOLOGY_CHANGED,
+        /** Data was truncated */
+        TRUNCATION,
+        /** Ordinary periodic reload */
+        PERIODIC,
+        /** Data was replayed either from the commit log or a batch log */
+        DATA_REPLAYED,
+        /** When repair task started */
+        REPAIR,
+        /** A request over forced by users to reload. */
+        USER_REQUESTED,
+    }
+
+    protected final TableMetadataRef metadata;
+    protected final Directories directories;
+    protected final Tracker dataTracker;
+
+    public StorageHandler(TableMetadataRef metadata, Directories directories, Tracker dataTracker)
+    {
+        Preconditions.checkNotNull(directories, "Directories should not be null");
+
+        this.metadata = metadata;
+        this.directories = directories;
+        this.dataTracker = dataTracker;
+    }
+
+    /**
+     * @return true if the node is ready to serve data for this table. This means that the
+     *         node is not bootstrapping and that no data may be missing, e.g. if sstables are
+     *         being downloaded from remote storage or streamed from other nodes then isReady()
+     *         would return false. Generally, user read queries should not succeed if this method
+     *         returns false.
+     */
+    public abstract boolean isReady();
+
+    /**
+     * Load the initial sstables into the tracker that was passed in to the constructor.
+     *
+     * @return the sstables that were loaded
+     */
+    public abstract Collection<SSTableReader> loadInitialSSTables();
+
+    /**
+     * Reload any sstables that may have been created and not yet loaded. This is normally
+     * a no-op for the default local storage, but for remote storage implementations it
+     * signals that sstables need to be refreshed.
+     */
+    public abstract void reloadSSTables(ReloadReason reason);
```

Comment:
for now, it's not needed..

Reply:
Yeah, I know, but since this is going to be a kind of a public API which should rather not fluctuate frequently, I was wondering if it makes sense for this methods to return the reloaded tables in the context of the future usages?

Reply:
added returned value: reloaded sstables

Reply:
@jacek-lewandowski Does it really make sense to require returning the collection of opened SSTables, which is never used anywhere? I understand it makes code to look nice, but I am not sure about semantical meaning of implementing it as there is no example of possible usage.